### PR TITLE
Handle out funcs with c10_full.

### PR DESCRIPTION
**Pull request by ailzhang**
https://github.com/pytorch/pytorch/pull/49013

### CODE REVIEW

File: scripts/gen.py [hunk 1]
```diff
@@ -638,16 +638,25 @@ def get_reference_param(params, fnopts=None):
...
+                     params,
+                     param_vars,
+                     ref_param,
+                     fnopts,
+                     c10_full_out=False):
   types = get_template_type_list(rtype)
   retstr = '{}('.format(rtype_str)
-  for i, ttype in enumerate(types):
+  n = len(types)
+  for i in range(n):
```
Comment (davidel) [inline scripts/gen.py:652]: We specify `range()` base explicitly all around our Python code:
```python
for i in range(0, n):
  ...
```

File: scripts/gen.py [hunk 2]
```diff
@@ -685,20 +694,28 @@ def generate_exit_debug_code(t, fname, rname, params, param_vars):
...
+                         ref_param,
+                         fnopts,
+                         c10_full_out=False):
   assert isinstance(t, lark.tree.Tree)
   rtype = t.children[0]
   ctype = type_core(rtype)
   if ctype == 'std::tuple':
     retstr = get_tuple_return(rtype, rtype_str, rname, params, param_vars,
-                              ref_param, fnopts)
+                              ref_param, fnopts, c10_full_out)
```
Comment (davidel) [inline scripts/gen.py:711]: `c10_full_out=c10_full_out`

File: scripts/gen.py [hunk 3]
```diff
@@ -715,8 +732,13 @@ def generate_result_assignment(t, rname):
...
 
-def get_handling_function(ctx, fname, xla_ref_param, param_vars):
-  function = _XLA_FUNCTIONS.get(fname, None) or ctx.get_function(fname)
+def get_handling_function(ctx,
+                          fname,
+                          xla_ref_param,
+                          param_vars,
+                          c10_full_out=False):
+  function = _XLA_FUNCTIONS.get(fname, None) or ctx.get_function(
+      fname, c10_full_out)
```
Comment (davidel) [inline scripts/gen.py:741]: `c10_full_out=c10_full_out`

File: scripts/gen.py [hunk 4]
```diff
@@ -785,6 +807,12 @@ def generate_outfn_result_copy(dest, src):
       dest, src)
 
 
+def is_c10_out(params):
+  ptype = param_type(params[0])
+  ctype = type_core(ptype)
+  return ctype != 'Tensor' or type_is_const(ptype)
```
Comment (davidel) [inline scripts/gen.py:813]: I'd a small comment outlining the two cases.

File: scripts/gen.py [hunk 5]
```diff
@@ -880,12 +909,14 @@ def generate_aten_to_xla(ctx, tree, rwxtree, fname, sig, rwsig, params, fnopts):
     code += '  at::TensorOptions options = at::TensorOptions().device(device).layout(layout).pinned_memory(pin_memory).dtype(dtype);\n'
     code += '  {}{};\n'.format(
         result_assign,
-        get_handling_function(ctx, fname, xla_ref_param, param_vars))
+        get_handling_function(ctx, fname, xla_ref_param, param_vars,
+                              c10_full_out))
```
Comment (davidel) [inline scripts/gen.py:913]: `c10_full_out=c10_full_out`

File: scripts/gen.py [hunk 6]
```diff
@@ -880,12 +909,14 @@ def generate_aten_to_xla(ctx, tree, rwxtree, fname, sig, rwsig, params, fnopts):
...
     code = code.replace(', '.join(target_options), 'options')
   else:
     code += '  {}{};\n'.format(
         result_assign,
-        get_handling_function(ctx, fname, xla_ref_param, param_vars))
+        get_handling_function(ctx, fname, xla_ref_param, param_vars,
+                              c10_full_out))
```
Comment (davidel) [inline scripts/gen.py:919]: `c10_full_out=c10_full_out`

File: scripts/gen.py [hunk 7]
```diff
@@ -895,7 +926,7 @@ def generate_aten_to_xla(ctx, tree, rwxtree, fname, sig, rwsig, params, fnopts):
                                    params, param_vars)
   code += generate_return_stmt(tree, get_return_type_str(rwxtree, rwsig), fname,
                                _RESULT_NAME if result_assign else None, params,
-                               param_vars, ref_param, fnopts)
+                               param_vars, ref_param, fnopts, c10_full_out)
```
Comment (davidel) [inline scripts/gen.py:929]: `c10_full_out=c10_full_out`

**Review by davidel — CHANGES_REQUESTED** [review]
I have to say ... it's changes like these w/out backward compatibility that can make C++ API users really unhappy 😄

File: scripts/gen.py (hunk 4 above)
Comment (ailzhang) [reply scripts/gen.py]: After talking to @smessmer I actually removed this check :D

**Review by davidel — APPROVED** [review]

**Pull request merged**